origin-notifications/app.js

```diff
@@ -0,0 +1,96 @@
+require('dotenv').config()
+
+const express = require('express')
+const bodyParser = require('body-parser')
+const webpush = require('web-push')
+const Sequelize = require('sequelize')
+
+const app = express()
+const port = 3456
+
+const PushSubscription = require('./src/models').PushSubscription
+const emailAddress = process.env.VAPID_EMAIL_ADDRESS
+const privateKey = process.env.VAPID_PRIVATE_KEY
+const publicKey = process.env.VAPID_PUBLIC_KEY
+
+webpush.setVapidDetails(
+  `mailto:${emailAddress}`,
+  publicKey,
+  privateKey
+)
+
+// should be tightened up for security
+app.use((req, res, next) => {
+  res.header("Access-Control-Allow-Origin", "*")
+  res.header("Access-Control-Allow-Headers", "Origin, X-Requested-With, Content-Type, Accept")
+
+  next()
+})
+
+app.use(bodyParser.json())
+
+// currently showing all subscriptions in an unauthenticated index view - maybe not a good idea in production?
+app.get('/', async (req, res) => {
+  const subs = await PushSubscription.findAll()
+
+  res.send(`<h1>${subs.length} Push Subscriptions</h1><ul>${subs.map(s => `<li><pre style="white-space: pre-wrap;word-break: break-all">${JSON.stringify(s)}</pre></li>`)}</ul>`)
+})
+
+app.post('/', async(req, res) => {
+  const { account, endpoint } = req.body
+  // cannot upsert because neither is unique
+  const existing = await PushSubscription.find({
+    where: { account, endpoint }
+  })
+
+  if (existing) {
+    return res.sendStatus(200)
+  }
+
+  await PushSubscription.create(req.body)
+
+  res.sendStatus(201)
+})
+
+app.post('/events', async (req, res) => {
```

Comment:
Need to add authentication otherwise user could get spammed by an external service calling this webhook.
We could have a secret that the listener passes as a header value in the request and that the notifier service checks. Another possibility would be to rely on internal queues rather than webhooks: listener would enqueue events and notifier would dequeue and process. Let's discuss...

Reply:
This doesn't need to be exposed to the outside world right? The simple solution here is I don't expose this endpoint outside the Kubernetes cluster. The listener can just call it from inside the cluster.

Reply:
Correct, @tomlinton 👌

Reply:
Excellent idea @tomlinton !